### PR TITLE
Fixes #10200: Change readme layout for Hooks

### PR DESCRIPTION
**Pull request by fanf**
https://www.rudder-project.org/redmine/issues/10200

### CODE REVIEW

File: rudder-core/src/main/resources/hooks.d/policy-generation-finished/readme.adoc [hunk 1]
```diff
@@ -0,0 +1,26 @@
+===== policy-generation-finished
+
+====== When/What ?
+
+This directory contains hooks executed after poplicies are fully
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-finished/readme.adoc]: policies

File: rudder-core/src/main/resources/hooks.d/policy-generation-finished/readme.adoc [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+generated for all nodes, and these new policies are available for
+download for the node.
+
+Typically, these hooks are used to log information about the
+generation which just happened or notify third parties that
+new policies are available (for ex: cf-serverd SIGHUP)
+
+====== Parameters
+
+Hooks parameter are passed by environment variable:
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-finished/readme.adoc]: parameters

File: rudder-core/src/main/resources/hooks.d/policy-generation-finished/readme.adoc [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+- ```RUDDER_GENERATION_DATETIME       ```: ISO-8601 YYYY-MM-ddTHH:mm:ss.sssZ
+date/time that identify that policy generation.
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-finished/readme.adoc]: identifies

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc [hunk 4]
```diff
@@ -0,0 +1,26 @@
+===== policy-generation-node-finished
+
+====== When/What ?
+
+This directory contains hooks executed after policies are fully
+generated for node and made available for the node to download.
+
+Typically, these hooks interact with external services using
+knowledged from the generated policies  (ex: send node-properties
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc]: knowledge

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+JSON file to a third party service).
+
+====== Parameters
+
+Hooks parameter are passed by environment variable:
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc]: parameters

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+- ```RUDDER_GENERATION_DATETIME```: generation datetime: ISO-8601 YYYY-MM-ddTHH:mm:ss.sssZ
+date/time that identify that policy generation start
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc]: identifies

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+- ```RUDDER_NODEID             ```: the nodeId
+- ```RUDDER_POLICIES_DIRECTORY ```: new policies directory (for ex for nodes
+under root: /var/rudder/share/$RUDDER_NODEID/rules)
+- ```RUDDER_AGENT_TYPE         ```: agent type ("cfengine-nova" or
+"cfengine-community")
+
+Technically, you could infer RUDDER_NEXT_POLICIES_DIRECTORY, from RUDDER_NODEID, but it's tedious
+for nodes begind relay, and it is just simpler to don't have to track what are the Rudder internal names,
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc]: behind a relay

Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc]: not to have to

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc [hunk 8]
```diff
@@ -0,0 +1,29 @@
+===== policy-generation-node-ready
+
+====== When/What ?
+
+This directory contains hooks executed after policies are fully
+generated for node, but before these new policies replace the old
+one (technically, before we move /var/rudder/share/$NODEID/rules.new
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc]: ones

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+to /var/rudder/share/$NODEID/rules).
+
+Typically, these hooks proceed to sanity checks on the new policies
+(ex: cf-promises), update permission on files, or interact with
+external services using knowledged from the generated policies
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc]: knowledge

Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc]: information is probably better here

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+- ```RUDDER_GENERATION_DATETIME    ```: generation datetime: ISO-8601
+YYYY-MM-ddTHH:mm:ss.sssZ date/time that identify that policy generation start
+- ```RUDDER_NODEID                 ```: the nodeId
+- ```RUDDER_NEXT_POLICIES_DIRECTORY```: new policies directory (for ex for nodes
+under root: /var/rudder/share/$RUDDER_NODEID/rules.new)
+- ```RUDDER_AGENT_TYPE             ```: agent type ("cfengine-nova" or
+"cfengine-community")
+
+Technically, you could infer RUDDER_NEXT_POLICIES_DIRECTORY, from RUDDER_NODEID, but it's tedious
+for nodes begind relay, and it is just simpler to don't have to track what are the Rudder internal names,
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc]: behind a relay

Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc]: not to have to

File: rudder-core/src/main/resources/hooks.d/policy-generation-started/readme.adoc [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+This directory contains hooks executed when a policy generation starts.
+
+Typically, these hooks are used to log information about the
+generation which just started or notify third parties that
+shared information to node should be updated (shared-folder, etc).
+
+====== Parameters
+
+Hooks parameter are passed by environment variable:
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-started/readme.adoc]: parameters

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 12]
```diff
@@ -0,0 +1,32 @@
+==== Generalities about hooks
+
+*Hooks are organized in subdirectories*. The root of the sub-directories os ```/opt/rudder/etc/hooks.d/```
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: is

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+
+Each sub-directory has a name related to the event that will trigger the hooks
+execution.
+By default, a hook directory contains a template example and a Readme.txt file
+explaining the generalities about hooks and the specificities of that hook (parameters, etc)
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: Missing . at the end of the sentence

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+
+*Hooks must be executable*. All executable file will be used as hooks, and
+non executable one will be ignore (which allow to put other file in these
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: ignored

Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: allows to put other files

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+directory, like a readme, for example).
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: directories

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+*Hooks parameter are passed by environment variable*. Rudder will fill dedicated
+environment variable for each hooks.
+
+*Hooks are executed sequentially, in lexical order*. We encourage
+you to use the patter "NN-hookname", with NN a number like
+"05", "20", etc.
+
+*Hooks have normalized returned code*. Return code on hooks are interpreted as follow:
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: return codes

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+- ```0     ```: success, no log (appart if debug one)          , continue to next hook
+- ```1-31  ```: error  , error   log in /var/log/rudder/webapp/, stop processing
+- ```32-63 ```: warning, warning log in /var/log/rudder/webapp/, continue to next hook
+- ```64-255```: reserved for futur use case. Behavior may change without notice.
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: future

**Comment by fanf** [comment]
PR rebased

**Comment by amousset** [comment]
Asciidoc uses a simple backquote for code.

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+YYYY-MM-ddTHH:mm:ss.sssZ date/time that identify that policy generation start
+- ```RUDDER_NODEID                 ```: the nodeId
+- ```RUDDER_NEXT_POLICIES_DIRECTORY```: new policies directory (for ex for nodes
+under root: /var/rudder/share/$RUDDER_NODEID/rules.new)
+- ```RUDDER_AGENT_TYPE             ```: agent type ("cfengine-nova" or
+"cfengine-community")
+
+Technically, you could infer RUDDER_NEXT_POLICIES_DIRECTORY, from RUDDER_NODEID, but it's tedious
+for nodes behind a relay, and it is just simpler not to have to track what are the Rudder internal names,
+which may change without notice.
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-ready/readme.adoc]: Could you add a newline here?

File: rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+date/time that identify that policy generation start
+- ```RUDDER_NODEID             ```: the nodeId
+- ```RUDDER_POLICIES_DIRECTORY ```: new policies directory (for ex for nodes
+under root: /var/rudder/share/$RUDDER_NODEID/rules)
+- ```RUDDER_AGENT_TYPE         ```: agent type ("cfengine-nova" or
+"cfengine-community")
+
+Technically, you could infer RUDDER_NEXT_POLICIES_DIRECTORY, from RUDDER_NODEID, but it's tedious
+for nodes behind a relay, and it is just simpler not to have to track what are the Rudder internal names,
+which may change without notice.
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/policy-generation-node-finished/readme.adoc]: Could you add a newline here?

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+- ```0     ```: success, no log (appart if debug one)          , continue to next hook
+- ```1-31  ```: error  , error   log in /var/log/rudder/webapp/, stop processing
+- ```32-63 ```: warning, warning log in /var/log/rudder/webapp/, continue to next hook
+- ```64-255```: reserved for future use case. Behavior may change without notice.
+
+
+== Available Hooked events
+
+For now, all hooks are related to different steps of the policy generation
+process. In the future, more hooks will be supported like node acceptation.
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: Could you add a newline here?

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+
+*Hooks have normalized returned code*. Return codes on hooks are interpreted as follow:
+
+- ```0     ```: success, no log (appart if debug one)          , continue to next hook
+- ```1-31  ```: error  , error   log in /var/log/rudder/webapp/, stop processing
+- ```32-63 ```: warning, warning log in /var/log/rudder/webapp/, continue to next hook
+- ```64-255```: reserved for future use case. Behavior may change without notice.
+
+
+== Available Hooked events
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: We should avoid this level of title, as we are limited to 5 in the doc. We should use == here for different hook types.

**Comment by fanf** [comment]
PR rebased

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 22]
```diff
@@ -0,0 +1,31 @@
+= Generalities about hooks
+
+*Hooks are organized in subdirectories*. The root of the sub-directories is ```/opt/rudder/etc/hooks.d/```
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: This should be simple backquote

File: rudder-core/src/main/resources/hooks.d/readme.adoc [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+*Hooks parameter are passed by environment variable*. Rudder will fill dedicated
+environment variable for each hooks.
+
+*Hooks are executed sequentially, in lexical order*. We encourage
+you to use the patter "NN-hookname", with NN a number like
+"05", "20", etc.
+
+*Hooks have normalized returned code*. Return codes on hooks are interpreted as follow:
+
+- `0     `: success, no log (appart if debug one)          , continue to next hook
```
Comment (amousset) [inline rudder-core/src/main/resources/hooks.d/readme.adoc]: Asciidoc does not understand spaces between backquotes :(

**Comment by fanf** [comment]
PR rebased

**Comment by fanf** [comment]
PR rebased

**Pull request merged**